core/chains/evm/logpoller/dsl.go

```diff
@@ -0,0 +1,239 @@
+
+	ubig "github.com/smartcontractkit/chainlink/v2/core/chains/evm/utils/big"
+)
+
+type ComparisonOperator int
+
+const (
+	Eq ComparisonOperator = iota
+	Neq
+	Gt
+	Lt
+	Gte
+	Lte
+)
+
+type SortDirection int
+
+const (
+	Asc SortDirection = iota
+	Desc
+)
+
+var (
+	DefaultSortAndLimit = SortAndLimit{
+		sortBy: []SortBy{
+			{field: "block_number", dir: Asc},
+			{field: "log_index", dir: Asc},
+		},
+		limit: 0,
+	}
+)
+
+type SortAndLimit struct {
+	sortBy []SortBy
+	limit  int
+}
+
+type SortBy struct {
+	field string
+	dir   SortDirection
+}
+
+func NewSortAndLimit(limit int, sortBy ...SortBy) SortAndLimit {
+	return SortAndLimit{sortBy: sortBy, limit: limit}
+}
+
+func NewSortBy(field string, dir SortDirection) SortBy {
+	return SortBy{field: field, dir: dir}
+}
+
+type QFilter interface {
+	accept(visitor Visitor)
+}
+
+type AndFilter struct {
+	filters []QFilter
+}
+
+func NewAndFilter(filters ...QFilter) *AndFilter {
+	return &AndFilter{filters: filters}
+}
+
+func NewBasicAndFilter(address common.Address, eventSig common.Hash, filters ...QFilter) *AndFilter {
+	allFilters := make([]QFilter, 0, len(filters)+2)
+	allFilters = append(allFilters, NewAddressFilter(address), NewEventSigFilter(eventSig))
+	allFilters = append(allFilters, filters...)
+	return NewAndFilter(allFilters...)
+}
+
+func AppendedNewFilter(root *AndFilter, other ...QFilter) *AndFilter {
+	filters := make([]QFilter, 0, len(root.filters)+len(other))
+	filters = append(filters, root.filters...)
+	filters = append(filters, other...)
+	return NewAndFilter(filters...)
+}
+
+func (f *AndFilter) accept(visitor Visitor) {
+	visitor.visitAndFilter(*f)
+}
+
+type EvmChainIdFilter struct {
+	chainId *ubig.Big
+}
+
+func NewEvmChainIdFilter(chainId *big.Int) *EvmChainIdFilter {
+	return &EvmChainIdFilter{chainId: ubig.New(chainId)}
+}
+
+func (f *EvmChainIdFilter) accept(visitor Visitor) {
+	visitor.visitEvmChainIdFilter(*f)
+}
+
+type AddressFilter struct {
+	address []common.Address
+}
+
+func NewAddressFilter(address ...common.Address) *AddressFilter {
+	return &AddressFilter{address: address}
+}
+
+func (f *AddressFilter) accept(visitor Visitor) {
+	visitor.visitAddressFilter(*f)
+}
+
+type EventSigFilter struct {
+	eventSig []common.Hash
+}
+
+func NewEventSigFilter(eventSig ...common.Hash) *EventSigFilter {
+	return &EventSigFilter{eventSig: eventSig}
+}
+
+func (f *EventSigFilter) accept(visitor Visitor) {
+	visitor.visitEventSigFilter(*f)
+}
+
+type DataWordFilter struct {
+	index    int
+	operator ComparisonOperator
+	value    common.Hash
+}
+
+func NewDataWordFilter(index int, operator ComparisonOperator, value common.Hash) *DataWordFilter {
+	return &DataWordFilter{index: index, operator: operator, value: value}
+}
+
+func NewDataWordGteFilter(index int, value common.Hash) *DataWordFilter {
+	return NewDataWordFilter(index, Gte, value)
+}
+
+func NewDataWordLteFilter(index int, value common.Hash) *DataWordFilter {
+	return NewDataWordFilter(index, Lte, value)
+}
```

Comment:
For consistency, maybe we should just have `NewDataWordFilter(index int, operator ComparisonOperator, value common.Hash)`?  Then it would be more similar to `NewTopicFilter`, `NewBlockFilter`, and `NewTimestamp`, and support all comparison operators instead of just two.

Reply:
I'm not sure if I follow, there is `NewDataWordFilter(index int, operator ComparisonOperator, value common.Hash)`. I've added `NewDataWordLteFilter` and `NewDataWordGteFilter` as convenient functions, because Lte and Gte are the most popular operators. However, user still can use any type of operator with a default constructor `NewDataWordFilter`